sherpa/data.py

```diff
@@ -866,20 +866,33 @@ class DataSimulFit(NoNewAttributesAfterInit):
 
     """
 
-    def __init__(self, name, datasets):
-        self.name = name
+    def __init__(self, name, datasets, numcores=1):
         if len(datasets) == 0:
             raise DataErr('zerodatasimulfit', type(self).__name__)
+        self.name = name
         self.datasets = tuple(datasets)
+        self.numcores = numcores
         NoNewAttributesAfterInit.__init__(self)
 
     def eval_model_to_fit(self, modelfuncs):
-        total_model = []
-
-        for func, data in zip(modelfuncs, self.datasets):
-            total_model.append(data.eval_model_to_fit(func))
-
-        return numpy.concatenate(total_model)
+        if self.numcores == 1:
+            total_model = []
+            for func, data in zip(modelfuncs, self.datasets):
+                tmp_model = data.eval_model_to_fit(func)
+                total_model.append(tmp_model)
+            return numpy.concatenate(total_model)
+        else:
+        # best to make this a different derived class
+            funcs = []
+            datasets = []
+            for func, data in zip(modelfuncs, self.datasets):
+                funcs.append(func)
+                datasets.append(data.get_indep(filter=False))
+            total_model = parallel_map_funcs(funcs, datasets)
```

Comment:
I'm sure you thought about it, but I'm curious why you don't use `eval_model_to_fit` in this case, something like
```
for ...
    funcs.append(data.eval_model_to_fit)
    arg_for_funcs.append(func)
total_model = parallel_map_funcs(funcs, arg_for_funcs)
```
Since the two branches for 1 or more cores are different, I suggest a small comment in the code.

Reply:
The original ```parallel_map``` assumes the arguments are numpy arrays to do its magic, which is not the case here.

I don't have data and not very familiar with what models are _time consuming_ so I relied on SDS giving me, the one test which goes from 3:20 to 3:14secs using 4 cores is the one that @anetasie gave me.

BTW, if one wants to have a runtime option of setting ```numcores``` then do tell how that can be done, enjoy.

Reply:
I'm thinking hard a about a model I can try for this, hopefully I get an idea over the weekend.

Reply:
BTW,  a good thing I looked at this comment cause the prototype for ```parallel_map_funcs``` is:
```
def parallel_map_funcs(funcs, datasets, numcores=None):
```

so the call to ```parallel_map_funcs``` as written, see comment above, was missing the ```numcores``` option so not sure how this ever worked, yikes!   Anyway, when I tried @hamogu's suggestion:

```
(sherpa) [dtn@devel12 sherpa]$ git diff sherpa/data.py
diff --git a/sherpa/data.py b/sherpa/data.py
index b9af0528..5c9faa6a 100644
--- a/sherpa/data.py
+++ b/sherpa/data.py
@@ -885,10 +885,14 @@ class DataSimulFit(NoNewAttributesAfterInit):
         # best to make this a different derived class
             funcs = []
             datasets = []
+            arg_for_funcs = []
             for func, data in zip(modelfuncs, self.datasets):
-                funcs.append(func)
-                datasets.append(data.get_indep(filter=False))
-            total_model = parallel_map_funcs(funcs, datasets)
+                funcs.append(data.eval_model_to_fit)
+                arg_for_funcs.append(func)
+            total_model = parallel_map_funcs(funcs, arg_for_funcs, self.numcores)
+            #     funcs.append(func)
+            #     datasets.append(data.get_indep(filter=False))
+            # total_model = parallel_map_funcs(funcs, datasets, self.numcores)
             all_model = []
             for model, data in zip(total_model, self.datasets):
                 all_model.append(data.apply_filter(model))
```
I get:
```
(sherpa) [dtn@devel12 PR483]$ fit_3_spectra.py 3
Process Process-2:
Traceback (most recent call last):
  File "/export/miniconda/envs/sherpa/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/export/miniconda/envs/sherpa/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/export/sherpa/sherpa/utils/__init__.py", line 2590, in worker
    out_q.put((ii, list(vals)))
  File "/export/sherpa/sherpa/astro/data.py", line 1663, in eval_model_to_fit
    return self.apply_filter(modelfunc(*self.get_indep(filter=True)))
  File "/export/sherpa/sherpa/astro/data.py", line 1221, in apply_filter
    elif len(data) != len(self.counts):
TypeError: object of type 'numpy.float64' has no len()
Process Process-3:
Traceback (most recent call last):
  File "/export/miniconda/envs/sherpa/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/export/miniconda/envs/sherpa/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/export/sherpa/sherpa/utils/__init__.py", line 2590, in worker
    out_q.put((ii, list(vals)))
  File "/export/sherpa/sherpa/astro/data.py", line 1663, in eval_model_to_fit
    return self.apply_filter(modelfunc(*self.get_indep(filter=True)))
  File "/export/sherpa/sherpa/astro/data.py", line 1221, in apply_filter
    elif len(data) != len(self.counts):
TypeError: object of type 'numpy.float64' has no len()
Process Process-4:
Traceback (most recent call last):
  File "/export/miniconda/envs/sherpa/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/export/miniconda/envs/sherpa/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "/export/sherpa/sherpa/utils/__init__.py", line 2590, in worker
    out_q.put((ii, list(vals)))
  File "/export/sherpa/sherpa/astro/data.py", line 1663, in eval_model_to_fit
    return self.apply_filter(modelfunc(*self.get_indep(filter=True)))
  File "/export/sherpa/sherpa/astro/data.py", line 1221, in apply_filter
    elif len(data) != len(self.counts):
TypeError: object of type 'numpy.float64' has no len()
Traceback (most recent call last):
  File "./fit_3_spectra.py.orig", line 136, in <module>
    tst(numcores)
  File "./fit_3_spectra.py.orig", line 127, in tst
    fit(1, 2, 3, numcores=numcores)
  File "<string>", line 1, in fit
  File "/export/sherpa/sherpa/astro/ui/utils.py", line 9907, in fit
    self._fit(id, *otherids, **kwargs)
  File "/export/sherpa/sherpa/astro/ui/utils.py", line 10013, in _fit
    res = f.fit(**kwargs)
  File "/export/sherpa/sherpa/fit.py", line 56, in run
    result = func(fit, *args, **kwargs)
  File "/export/sherpa/sherpa/fit.py", line 1231, in fit
    init_stat = self.calc_stat()
  File "/export/sherpa/sherpa/fit.py", line 1116, in calc_stat
    return self._calc_stat()[0]
  File "/export/sherpa/sherpa/fit.py", line 1097, in _calc_stat
    return self.stat.calc_stat(self.data, self.model)
  File "/export/sherpa/sherpa/stats/__init__.py", line 916, in calc_stat
    data_model = data.eval_model_to_fit(model)
  File "/export/sherpa/sherpa/data.py", line 892, in eval_model_to_fit
    total_model = parallel_map_funcs(funcs, arg_for_funcs, self.numcores)
  File "/export/sherpa/sherpa/utils/__init__.py", line 2869, in parallel_map_funcs
    return run_tasks(procs, err_q, out_q, numcores)
  File "/export/sherpa/sherpa/utils/__init__.py", line 2623, in run_tasks
    vals.extend(r)
TypeError: 'NoneType' object is not iterable
```
 Will stick with the original code, apart from the additional call to pass ```numcores```.